File: mf2/java/src/main/java/com/box/l10n/mojito/mf2/Mf2ModelValidation.java
package com.box.l10n.mojito.mf2;

import java.util.List;
import java.util.Map;

/** Structural checks that Java's model types alone cannot express. */
final class Mf2ModelValidation {
    private Mf2ModelValidation() {}

    static void validate(Mf2Message message) throws Mf2Exception {
        require(message != null);
        for (var declaration : message.declarations()) {
            require(declaration.name() != null);
            expression(declaration.value());
        }
        switch (message) {
            case Mf2Message.PatternMessage value -> pattern(value.pattern());
            case Mf2Message.SelectMessage value -> {
                for (var selector : value.selectors()) require(selector.name() != null);
                for (var variant : value.variants()) {
                    for (var key : variant.keys()) {
                        if (key instanceof Mf2Message.LiteralVariantKey literal) require(literal.value() != null);
                    }
                    pattern(variant.value());
                }
            }
        }
    }

    private static void expression(Mf2Message.Expression expression) throws Mf2Exception {
        require(expression != null);
        require(expression.arg() != null || expression.function() != null);
        if (expression.arg() != null) argument(expression.arg());
        if (expression.function() != null) {
            require(expression.function().name() != null);
            options(expression.function().options());
        }
        attributes(expression.attributes());
    }

    private static void argument(Mf2Message.ExpressionArgument argument) throws Mf2Exception {
        require(argument != null);
        switch (argument) {
            case Mf2Message.LiteralArgument literal -> require(literal.value() != null);
            case Mf2Message.VariableArgument variable -> require(variable.name() != null);
        }
    }

    private static void options(Map<String, Mf2Message.ExpressionArgument> options) throws Mf2Exception {
        for (var value : options.values()) argument(value);
    }

    private static void attributes(Map<String, Mf2Message.AttributeValue> attributes) throws Mf2Exception {
        for (var value : attributes.values()) {
            switch (value) {
                case Mf2Message.PresentAttribute present -> require(present.value());
                case Mf2Message.LiteralAttribute literal -> {
                    require(literal.value() instanceof Mf2Message.LiteralArgument);
                    argument(literal.value());
                }
            }
        }
    }

    private static void pattern(List<Mf2Message.PatternPart> pattern) throws Mf2Exception {
        for (var part : pattern) {
            switch (part) {
                case Mf2Message.TextPart text -> require(text.value() != null);
                case Mf2Message.ExpressionPart value -> expression(value.expression());
                case Mf2Message.MarkupPart value -> {
                    var markup = value.markup();
                    require(markup != null && markup.kind() != null && markup.name() != null);
                    options(markup.options());
                    attributes(markup.attributes());
                }
            }
        }
    }

    private static void require(boolean valid) throws Mf2Exception {
        if (!valid) throw new Mf2Exception("invalid-model", "Message model does not match the shared model schema.");
    }
}

File: mf2/kotlin/src/main/kotlin/com/box/l10n/mojito/mf2/Mf2ModelValidation.kt
package com.box.l10n.mojito.mf2

/** Validate the semantic fields of the shared model schema, ignoring extension properties. */
internal fun validateModelShape(model: Mf2Model) {
    fun invalid(): Nothing = throw Mf2Error("invalid-model", "Message model does not match the shared model schema.")
    fun obj(value: Any?): Map<*, *> = (value as? Map<*, *>)?.takeIf { it.keys.all { key -> key is String } } ?: invalid()
    fun list(value: Any?): List<*> = value as? List<*> ?: invalid()
    fun text(value: Any?) { if (value !is String) invalid() }
    fun argument(value: Any?, literalOnly: Boolean = false) {
        val item = obj(value)
        when (item["type"]) {
            "literal" -> text(item["value"])
            "variable" -> if (literalOnly) invalid() else text(item["name"])
            else -> invalid()
        }
    }
    fun fields(item: Map<*, *>) {
        if (item.containsKey("options")) obj(item["options"]).values.forEach { argument(it) }
        if (item.containsKey("attributes")) obj(item["attributes"]).values.forEach { if (it != true) argument(it, true) }
    }
    fun expression(value: Any?) {
        val item = obj(value)
        if (item["type"] != "expression" || (!item.containsKey("arg") && !item.containsKey("function"))) invalid()
        if (item.containsKey("arg")) argument(item["arg"])
        if (item.containsKey("function")) {
            val function = obj(item["function"])
            if (function["type"] != "function") invalid()
            text(function["name"])
            if (function.containsKey("options")) obj(function["options"]).values.forEach { argument(it) }
        }
        if (item.containsKey("attributes")) obj(item["attributes"]).values.forEach { if (it != true) argument(it, true) }
    }
    fun pattern(value: Any?) {
        for (part in list(value)) {
            if (part is String) continue
            val item = obj(part)
            when (item["type"]) {
                "expression" -> expression(item)
                "markup" -> { text(item["kind"]); text(item["name"]); fields(item) }
                else -> invalid()
            }
        }
    }
    for (raw in list(model["declarations"])) {
        val declaration = obj(raw)
        if (declaration["type"] !in setOf("input", "local")) invalid()
        text(declaration["name"])
        expression(declaration["value"])
    }
    when (model["type"]) {
        "message" -> pattern(model["pattern"])
        "select" -> {
            for (raw in list(model["selectors"])) {
                val selector = obj(raw)
                if (selector["type"] != "variable") invalid()
                text(selector["name"])
            }
            for (raw in list(model["variants"])) {
                val variant = obj(raw)
                for (key in list(variant["keys"])) {
                    val item = obj(key)
                    if (item["type"] == "*") { if (item.containsKey("value")) text(item["value"]) } else argument(item, true)
                }
                pattern(variant["value"])
            }
        }
        else -> invalid()
    }
}

internal fun detachedModelFields(fields: Map<String, Any?>): Map<String, Any?> =
    fields.mapValues { (_, value) -> if (value is Map<*, *>) value.toMap() else value }
